src/internal/pfsdb/commits.go

```diff
@@ -485,6 +502,71 @@ func getCommitChildren(ctx context.Context, extCtx sqlx.ExtContext, parentCommit
 	return children, nil
 }
 
+// CommitAncestry models a lineage of the CommitID values of the ancestors of Start.
+type CommitAncestry struct {
+	Start              CommitID
+	Lineage            map[CommitID]CommitID
+	EarliestDiscovered CommitID // EarliestDiscovered is the root of the commit ancestry tree if FoundRoot is true.
+	FoundRoot          bool     // FoundRoot is true if GetCommitAncestry reaches a root of a commit ancestry tree.
+}
+
+// GetCommitAncestry returns a CommitAncestry from startId up to maxDepth.
+func GetCommitAncestry(ctx context.Context, extCtx sqlx.ExtContext, startId CommitID, maxDepth uint) (*CommitAncestry, error) {
+	ancestry := &CommitAncestry{
+		Start:     startId,
+		Lineage:   make(map[CommitID]CommitID),
+		FoundRoot: true,
+	}
+	if err := ForEachCommitAncestor(ctx, extCtx, startId, maxDepth, func(parentId, childId CommitID) error {
+		ancestry.Lineage[parentId] = childId
+		ancestry.EarliestDiscovered = parentId
+		return nil
+	}); err != nil {
+		if errors.As(err, new(*MaxDepthReachedError)) {
+			ancestry.FoundRoot = false
+		} else {
+			return nil, errors.Wrap(err, "get commit ancestry")
+		}
+	}
+	return ancestry, nil
+}
+
+// ForEachCommitAncestor queries postgres for ancestors of startId up to the maxDepth. cb() is called for each ancestor.
+// maxDepth is optional.
+func ForEachCommitAncestor(ctx context.Context, extCtx sqlx.ExtContext, startId CommitID, maxDepth uint, cb func(parentId, childId CommitID) error) error {
+	if maxDepth == 0 {
+		maxDepth = DefaultMaxSearchDepth
+	}
+	query := `
+		WITH RECURSIVE ancestry AS (
+			SELECT parent, child, 0 as depth FROM pfs.commit_ancestry WHERE child = $1
+			UNION
+			SELECT ca.parent, ca.child, a.depth+1 FROM pfs.commit_ancestry ca
+			JOIN ancestry a ON ca.child = a.parent
+		)
+		SELECT a.parent, a.child, depth
+		FROM ancestry a WHERE depth <= $2;`
+	rows, err := extCtx.QueryContext(ctx, query, startId, maxDepth)
+	if err != nil {
+		return errors.Wrap(err, "get oldest commit ancestor")
+	}
+	defer rows.Close()
+	for rows.Next() {
+		var parent, child CommitID
+		var depth uint
+		if err := rows.Scan(&parent, &child, &depth); err != nil {
+			return errors.Wrap(err, "scanning parent and child row")
+		}
+		if err := cb(parent, child); err != nil {
+			return errors.Wrap(err, "calling cb() on parent and child")
+		}
+		if depth >= maxDepth {
```

Comment:
This seems unnecessary. The error type, and checking for this here. If the depth is handled correctly in the query, then we should always get the correct number of commits.

Reply:
This allows a caller to know that the max depth was reached as opposed to hitting the root of the ancestry tree. It's used by GetCommitAncestry to differentiate between the two outcomes.

Reply:
I don't see the value of that distinction, or having a specific type for the ancestry. It seems like getting the commit ancestry should just return a list of commits.

Reply:
The distinction saves a caller from having to call the function two times otherwise, there is no way to know that they've retrieved the ID of the root of the ancestry tree.